src/raftstore/coprocessor/split_check/size.rs

```diff
@@ -0,0 +1,255 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rocksdb::DB;
+use raftstore::store::{util, Msg};
+use util::transport::{RetryableSendCh, Sender};
+
+use super::super::{Coprocessor, ObserverContext, RegionObserver};
+use super::super::metrics::*;
+use super::Status;
+
+#[derive(Default)]
+pub struct SizeStatus {
+    current_size: u64,
+    split_key: Option<Vec<u8>>,
+}
+
+pub struct SizeCheckObserver<C> {
+    region_max_size: u64,
+    split_size: u64,
+    ch: RetryableSendCh<Msg, C>,
+}
+
+impl<C: Sender<Msg>> SizeCheckObserver<C> {
+    pub fn new(
+        region_max_size: u64,
+        split_size: u64,
+        ch: RetryableSendCh<Msg, C>,
+    ) -> SizeCheckObserver<C> {
+        SizeCheckObserver {
+            region_max_size,
+            split_size,
+            ch,
+        }
+    }
+}
+
+impl<C> Coprocessor for SizeCheckObserver<C> {}
+
+impl<C: Sender<Msg> + Send> RegionObserver for SizeCheckObserver<C> {
+    fn new_split_check_status(&self, ctx: &mut ObserverContext, status: &mut Status, engine: &DB) {
+        let size_status = SizeStatus::default();
+        let region = ctx.region();
+        let region_id = region.get_id();
+        let region_size = match util::get_region_approximate_size(engine, region) {
+            Ok(size) => size,
+            Err(e) => {
+                error!(
+                    "[region {}] failed to get approximate size: {}",
+                    region_id,
+                    e
+                );
+                // Need to check size.
+                status.size = Some(size_status);
+                return;
+            }
+        };
+
+        let res = Msg::ApproximateRegionSize {
+            region_id: region_id,
+            region_size: region_size,
+        };
+        if let Err(e) = self.ch.try_send(res) {
+            error!(
+                "[region {}] failed to send approximate region size: {}",
+                region_id,
+                e
+            );
+        }
+
+        REGION_SIZE_HISTOGRAM.observe(region_size as f64);
+        if region_size >= self.region_max_size {
+            info!(
+                "[region {}] approximate size {} >= {}, need to do split check",
+                region.get_id(),
+                region_size,
+                self.region_max_size
+            );
+            // Need to check size.
+            status.size = Some(size_status);
+        } // else { Does not need to check size. }
+    }
+
+    fn on_split_check(
+        &self,
+        _: &mut ObserverContext,
+        status: &mut Status,
+        key: &[u8],
+        value_size: u64,
+    ) -> Option<Vec<u8>> {
+        if let Some(size_status) = status.size.as_mut() {
+            size_status.current_size += key.len() as u64 + value_size;
+            if size_status.current_size > self.split_size && size_status.split_key.is_none() {
+                size_status.split_key = Some(key.to_vec());
```

Comment:
seem you don't need to clone the key every time, you can do it only when first current_size >= region_max_size